web/src/components/storage/PartitionsField.jsx

```diff
@@ -603,15 +637,69 @@ const Advanced = ({
   onBootChange,
   isLoading
 }) => {
-  const rootVolume = (volumes || []).find((i) => i.mountPath === "/");
+  const [isVolumeDialogOpen, setIsVolumeDialogOpen] = useState(false);
+  /** @type {[Volume|undefined, (volume: Volume) => void]} */
+  const [template, setTemplate] = useState();
+
+  const openVolumeDialog = () => setIsVolumeDialogOpen(true);
+
+  const closeVolumeDialog = () => setIsVolumeDialogOpen(false);
+
+  /** @type {(volume: Volume) => void} */
+  const onAcceptVolumeDialog = (volume) => {
+    closeVolumeDialog();
 
-  const addVolume = (volume) => onVolumesChange([...volumes, volume]);
+    const index = volumes.findIndex(v => v.mountPath === volume.mountPath);
+
+    if (index !== -1) {
+      const newVolumes = [...volumes];
+      newVolumes[index] = volume;
+      onVolumesChange(newVolumes);
+    } else {
+      onVolumesChange([...volumes, volume]);
+    }
+  };
 
   const resetVolumes = () => onVolumesChange([]);
 
-  const changeBtrfsSnapshots = ({ active }) => {
-    //  const rootVolume = volumes.find((i) => i.mountPath === "/");
+  /** @type {(mountPath: string) => void} */
+  const addVolume = (mountPath) => {
+    const template = templates.find(t => t.mountPath === mountPath);
+    setTemplate(template);
+    openVolumeDialog();
+  };
+
+  /**
+   * Possible mount paths to add.
+   * @type {() => string[]}
+   */
+  const mountPathOptions = () => {
+    const mountPaths = volumes.map(v => v.mountPath);
+    const isTransactional = isTransactionalSystem(templates);
+
+    return templates
+      .map(t => t.mountPath)
+      .filter(p => !mountPaths.includes(p))
+      .filter(p => !isTransactional || p.length);
+  };
+
+  /**
+   * Whether to show the button for adding a volume.
+   * @type {() => boolean}
+   */
+  const showAddVolume = () => {
+    const hasOptionalVolumes = () => {
+      return templates.find(t => t.mountPath.length && !t.outline.required) !== undefined;
+    };
+
+    return !isTransactionalSystem(templates) || hasOptionalVolumes();
+  };
```

Comment:
You already explained to me the reason why you tend to use functions for a better logic encapsulation. But still, I think that sometimes we're abusing a bit of their usage.

Here, I can see the benefit of having `hasOptionalVolumes` defined as an small function in `showAddVolume` function. It prevents to evaluate the `find` if it's actually not needed because is not a transactional system. But maybe `hasOptionalVolumes` also deserves a place in storage/utils and should recive templates as an argument instead of getting them from the scope, as `isTransactionalSysmte` alrady does. Thus, the `showAddVolume` would be not needed at all, and instead you can go for a `<If condition={!isTransactionalSystem(templates) || hasOptionalVolumes(templates)} ... />` below. Much clear and less indirection IMHO. Plus, developers can reuse `hasOptionalVolumes` when needed by taking a look at what we have in storage/utils, avoiding to create yet another function to do the same.

Anyway, I know that this is _just a matter of taste_.

Reply:
It sound good. Feel free to move it.

Reply:
The PR is already merged. Let's move it to utils in the next opportunity. Making a note in my todo list.